runtime/src/main/java/org/corfudb/protocols/wireprotocol/MsgHandlingFilter.java

```diff
@@ -0,0 +1,222 @@
+package org.corfudb.protocols.wireprotocol;
+
+import lombok.Data;
+import lombok.NonNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Representation of filtering function, it is composed of three generic actions each
+ * of which will be applied using instances of functional methods. Consequently,
+ * the input message will be transformed through the following steps.
+ *
+ * 1) PRE PROCESS: this function will be applied to a msg before processing the message.
+ *    an example is injecting a message before particular messages.
+ * 2) PROCESS: this function will be applied to a msg. Examples of such filtering
+ *    of a messages vary from allowing a message, to dropping it, or transforming
+ *    the content of message.
+ * 3) POST PROCESS: a function that will be applied at the end of transforming a message.
+ *    An example is injecting a message after processing certain messages.
+ *
+ */
+@Data
+public class MsgHandlingFilter implements Function<CorfuMsg, List<CorfuMsg>> {
```

Comment:
Overall a very nice class with lots of documentation !!! Great job and a great start!!
One overall question is that this is only for inbound messages. What about the outbound messages. 
I understand all use cases can be functionally simulated with filtering in one direction, but sometime there are implementation bugs that might only show up when you tamper outbound messages.

Reply:
Thanks @medhavidhawan . Yes, this commit was limited to inbound messages on client. I’ll add the others in the following commits.